Add mapping from algo OID to x509 SignatureAlgorithm for DSAWithSHA1

matthias50 opened the pull request:
This is preventing me from using this library to validate pkcs7 formatted AWS instance identity documents. See https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html

matthias50 committed:
Add mapping from algo OID to x509 SignatureAlgorithm for DSAWithSHA1

matthias50 commented:
@mastahyeti Can you please review and merge this if it looks good? I'd also like to get my other PR merged, but this one is more important as it is a blocker for me.

btoews commented:
First of all, sorry for the slow response. I have been traveling for work and haven't had time to look at your PRs until today. As for this change, DSA keys are widely being deprecated since 1024 bit keys are no longer considered safe. This is why I chose not to support DSA keys in this library. That said, you aren't always in control of what algorithms the signer is using. Maybe it makes sense to allow for verification with DSA certs but not signing? I believe that is the effect of your pull request, right?. Can you maybe add some tests to this PR, verifying that 1) DSA signature verification works and 2) that DSA signatures cannot be created?

matthias50 committed:
Added test case to verify validation pkcs7 doc signed with DSAWithSHA1

matthias50 commented:
> First of all, sorry for the slow response. I have been traveling for work and haven't had time to look at your PRs until today. As for this change, DSA keys are widely being deprecated since 1024 bit keys are no longer considered safe. This is why I chose not to support DSA keys in this library. That said, you aren't always in control of what algorithms the signer is using. Maybe it makes sense to allow for verification with DSA certs but not signing? I believe that is the effect of your pull request, right?. Can you maybe add some tests to this PR, verifying that 1) DSA signature verification works and 2) that DSA signatures cannot be created?

Thanks for the review.  Regarding 1) I added a test case which verifies DSA signature verification works using a CA and pkcs7 doc I generated using openssl. 

For 2) The map which I modified, `SignatureAlgorithmToX509SignatureAlgorithm` is only referenced one place in the code, `X509SignatureAlgorithm()` in `protocol.go`. `X509SignatureAlgorithm()` is only called by `verify()` in `verify.go`, so yes, this has nothing to do with signing and only has the effect of allowing allowing the code to map from a string to an algorithm using for signatures.

I tried to create a test case to validate that the DSA signatures could not be created, but ran into issue. The fakeca requires that the keys implement the crypto.Signer interface. This is fine for `crypt/ecdsa`, but `crypt/dsa` doesn't implement this interface. So I'm wondering how to even construct the code to try this.

In a related issue `Sign()` requires that the key be of type `crypto.Signer`. So basically, it seems like the fact that `crypt/dsa` doesn't implement the crypto.Signer interface makes it impossible to even try to create signatures with it.

Based on this, is just the test for 1) sufficient?

btoews commented on sign_test.go:
```diff
@@ -134,3 +135,86 @@ func TestSignRemoveHeaders(t *testing.T) {
 	}
 }
+
+func TestVerifyDSAWithSHA1(t *testing.T) {
```
This should probably go in `verify_test.go`. Ideally, we would generate the key/sig at runtime, but I guess that isn't easy since DSA doesn't implement `crypto.Signer`

https://github.com/mastahyeti/cms/blob/119fef08a078a262d54a40366d1c3f7de2e9d3f1/main_test.go#L23-L41

matthias50 replied:
Yea, that test was in the wrong spot. I moved it to verify_test.go. I agree that ideally we would generate the key/sig at runtime, but the fact that DSA doesn't implement crypto.Signer makes this a really big task. From what I can tell, it would require overhauling fakeca so that it wouldn't require crypto.Signer.

btoews replied:
That, or make another type that wraps `crypto/dsa.PrivateKey` and implements `crypto.Signer`. I might look into that later, but it isn't necessary for this PR.

btoews commented:
Thanks for the PR!

btoews commented:
I cut a [v0.0.7](https://github.com/mastahyeti/cms/releases/tag/v0.0.7) tag with this change.
